File: windows/net/samples/CameraToDVD/MediaState.cs
using System;
using System.Collections.Generic;
using System.Text;
using DirectShowLib;
using PrimoSoftware.DVDBuilder.VR;
using System.Diagnostics;

namespace CameraToDVD
{
    class MediaState
    {
        public MediaState()
        {
        }

        public IMediaControl mediaControl;
        //public IGraphBuilder graph;
        public IFilterGraph2 graph;
        public ICaptureGraphBuilder2 captureGraph;

        public IBaseFilter audioInput;
        public IBaseFilter videoInput;
        public IBaseFilter smartTee;
        public IBaseFilter previewRenderer;

        public IBaseFilter audioSampleGrabber;
        public IBaseFilter videoSampleGrabber;
        public ISampleGrabber audioGrabber;
        public ISampleGrabber videoGrabber;

        public IBaseFilter audioNullRenderer;
        public IBaseFilter videoNullRenderer;

        public DsROTEntry rot;

        public AMMediaType videoType = new AMMediaType();

        public IAMDroppedFrames droppedFrames;

        public PrimoSoftware.AVBlocks.Transcoder mpeg2Enc;
        public VideoRecorder dvdRecorder;

        public void Reset(bool full)
        {
            /*
             * NOTE:
             * Interfaces obtained with a cast (as) should not be released with Marshal.ReleaseComObject.
             */

            if (mediaControl != null)
            {
                mediaControl.StopWhenReady();
                mediaControl = null;
            }

            if (full && rot != null)
            {
                rot.Dispose();
                rot = null;
            }

            DsUtils.FreeAMMediaType(videoType);

            Util.DisposeObject(ref mpeg2Enc);

            if (previewRenderer != null)
            {
                IVideoWindow window = graph as IVideoWindow;
                window.put_Owner(IntPtr.Zero);
                window.put_Visible(OABool.False);
            }
            
            if (videoInput != null)
            {
                Util.NukeDownstream(graph, videoInput);
            }

            if (audioInput != null)
            {
                Util.NukeDownstream(graph, audioInput);
            }
            
            audioNullRenderer = null;
            videoNullRenderer = null;
            audioGrabber = null;
            videoGrabber = null;
            audioSampleGrabber = null;
            videoSampleGrabber = null;
            droppedFrames = null;
            previewRenderer = null;
            smartTee = null;

            if (full)
            {
                Util.ReleaseComObject(ref videoInput);
                Util.ReleaseComObject(ref audioInput);
                Util.ReleaseComObject(ref graph);
                Util.ReleaseComObject(ref captureGraph);
            }
            
            if (dvdRecorder != null)
            {
                bool res = dvdRecorder.Stop();
                if (!res)
                {
                    Trace.WriteLine("VideoRecorder.Stop() failed");

                    Util.CheckRecorderDevices(dvdRecorder);
                }

                for (int i = 0; i < dvdRecorder.Devices.Count; i++)
                {
                    if (dvdRecorder.GetDeviceError(i).Facility == PrimoSoftware.DVDBuilder.ErrorFacility.Success)
                        dvdRecorder.Devices[i].NotifyOSFileSystemChanged();
                    dvdRecorder.Devices[i].Dispose();
                }
            }

            Util.DisposeObject(ref dvdRecorder);
        }

        
    }
}

File: windows/net/samples.vb/CameraToDVD/MediaState.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DirectShowLib
Imports PrimoSoftware.DVDBuilder.VR
Imports System.Diagnostics

Namespace CameraToDVD
    Friend Class MediaState
        Public Sub New()
        End Sub

        Public mediaControl As IMediaControl
        'public IGraphBuilder graph;
        Public graph As IFilterGraph2
        Public captureGraph As ICaptureGraphBuilder2

        Public audioInput As IBaseFilter
        Public videoInput As IBaseFilter
        Public smartTee As IBaseFilter
        Public previewRenderer As IBaseFilter

        Public audioSampleGrabber As IBaseFilter
        Public videoSampleGrabber As IBaseFilter
        Public audioGrabber As ISampleGrabber
        Public videoGrabber As ISampleGrabber

        Public audioNullRenderer As IBaseFilter
        Public videoNullRenderer As IBaseFilter

        Public rot As DsROTEntry

        Public videoType As New AMMediaType()

        Public droppedFrames As IAMDroppedFrames

        Public mpeg2Enc As PrimoSoftware.AVBlocks.Transcoder
        Public dvdRecorder As VideoRecorder

        Public Sub Reset(ByVal full As Boolean)
            '            
            '             * NOTE:
            '             * Interfaces obtained with a cast (as) should not be released with Marshal.ReleaseComObject.
            '             

            If mediaControl IsNot Nothing Then
                mediaControl.StopWhenReady()
                mediaControl = Nothing
            End If

            If full AndAlso rot IsNot Nothing Then
                rot.Dispose()
                rot = Nothing
            End If

            DsUtils.FreeAMMediaType(videoType)

            Util.DisposeObject(mpeg2Enc)

            If previewRenderer IsNot Nothing Then
                Dim window As IVideoWindow = TryCast(graph, IVideoWindow)
                window.put_Owner(IntPtr.Zero)
                window.put_Visible(OABool.False)
            End If

            If videoInput IsNot Nothing Then
                Util.NukeDownstream(graph, videoInput)
            End If

            If audioInput IsNot Nothing Then
                Util.NukeDownstream(graph, audioInput)
            End If

            audioNullRenderer = Nothing
            videoNullRenderer = Nothing
            audioGrabber = Nothing
            videoGrabber = Nothing
            audioSampleGrabber = Nothing
            videoSampleGrabber = Nothing
            droppedFrames = Nothing
            previewRenderer = Nothing
            smartTee = Nothing

            If full Then
                Util.ReleaseComObject(videoInput)
                Util.ReleaseComObject(audioInput)
                Util.ReleaseComObject(graph)
                Util.ReleaseComObject(captureGraph)
            End If

            If dvdRecorder IsNot Nothing Then
                Dim res As Boolean = dvdRecorder.Stop()
                If (Not res) Then
                    Trace.WriteLine("VideoRecorder.Stop() failed")

                    Util.CheckRecorderDevices(dvdRecorder)
                End If

                For i As Integer = 0 To dvdRecorder.Devices.Count - 1
                    If dvdRecorder.GetDeviceError(i).Facility = PrimoSoftware.DVDBuilder.ErrorFacility.Success Then
                        dvdRecorder.Devices(i).NotifyOSFileSystemChanged()
                    End If
                    dvdRecorder.Devices(i).Dispose()
                Next i
            End If

            Util.DisposeObject(dvdRecorder)
        End Sub


    End Class
End Namespace
